NFS/nfsd/nfsd.8
.\"
.\" Copyright (c) 1999-2007 Apple Inc.  All rights reserved.
.\"
.\" @APPLE_LICENSE_HEADER_START@
.\" 
.\" This file contains Original Code and/or Modifications of Original Code
.\" as defined in and that are subject to the Apple Public Source License
.\" Version 2.0 (the 'License'). You may not use this file except in
.\" compliance with the License. Please obtain a copy of the License at
.\" http://www.opensource.apple.com/apsl/ and read it before using this
.\" file.
.\" 
.\" The Original Code and all software distributed under the License are
.\" distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
.\" EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
.\" INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
.\" Please see the License for the specific language governing rights and
.\" limitations under the License.
.\" 
.\" @APPLE_LICENSE_HEADER_END@
.\"
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)nfsd.8	8.4 (Berkeley) 3/29/95
.\"
.Dd November 10, 2006
.Dt NFSD 8
.Os
.Sh NAME
.Nm nfsd
.Nd 
.Tn NFS
server daemon
.Sh SYNOPSIS
.Nm nfsd
.Op Cm command
.Nm
.Op Fl F Ar exports_file
.Cm checkexports
.Nm
.Op Fl NRrtuv
.Op Fl F Ar exports_file
.Op Fl n Ar num_servers
.Op Fl p Ar nfsport
.Op Fl P Ar mountport
.Op Cm command
.Sh DESCRIPTION
.Nm nfsd
runs on an NFS server machine to service
.Tn NFS
and
.Tn MOUNT
protocol requests from NFS client machines.  In order for a machine
to act as an NFS server an NFS exports file,
.Pa /etc/exports ,
must exist and the
.Nm
service must be enabled.
.Pp
.Nm
listens for NFS service requests at the port indicated in the
.Tn NFS
server specification (2049) and the mountd thread listens for MOUNT
service requests at an available port registered with
.Xr portmap 8 .
.Pp
For more information on the NFS and MOUNT protocols see
.%T "Network File System Protocol Specification" ,
RFC1094 and
.%T "NFS: Network File System Version 3 Protocol Specification" .
.Pp
The
.Nm
daemon is a multi-threaded process that includes a number of threads
processing NFS requests, a thread to accept new socket connections and
a thread that processes NFS MOUNT protocol requests.
.Pp
When
.Nm
is started, it loads the export host addresses and options into the
kernel using the
.Xr nfssvc 2
system call.  After changing the list of exports (either directly or
indirectly via a change in netgroup membership), a hangup signal should
be sent to the daemon to get it to reload the export information.  This
can be accomplished with the
.Cm update
command described below.
.Pp
Any errors encountered while processing the export entries will
be logged via
.Xr syslog 3 .
.Pp
.Nm
is normally launched by
.Xr launchd 8 ;
however, the
.Nm
command may also be used to manipulate the service
using the following commands:
.Pp
.Bl -tag -width disable
.It Cm enable
Enables the
.Nm
service.
.It Cm disable
Disables the
.Nm
service.
.It Cm start
Starts the
.Nm
service.  Note: if the service is disabled it will not be restarted
on reboot.  Use the
.Cm enable
command to make the change permanent.
.It Cm stop
Stops the
.Nm
service.  Note: if the service is enabled it will be restarted on
reboot.  Use the
.Cm disable
command to make the change permanent.
.It Cm restart
Restarts the
.Nm
service (by stopping the service - it will restart automatically if
the
.Pa /etc/exports
file exists).
.It Cm update
Sends a SIGHUP to the running
.Nm
daemon to cause it to update its configuration.
.It Cm status
Displays whether the
.Nm
service is enabled and whether the
.Nm
daemon is currently running.
.It Cm checkexports
Checks the exports file and reports any errors (to stderr).  Note that
this can be useful to verify the validity of an alternate exports file
(using the
.Fl F
option below) prior to putting the changes in place and updating
.Nm .
.It Cm verbose [up|down]+
Sends USR1 and/or USR2 signal(s) to the running
.Nm
daemon to cause it to increase and/or decrease its logging level.  (See
below for details about
.Nm Ns 's
logging).
.El
.Pp
Logging is performed via
.Xr syslog 3
using the LOG_DAEMON facility.  By default, only messages up to priority
LOG_WARNING are logged.  Setting the verbose level to one will add
LOG_NOTICE messages which includes logging failed mount attempts.  A
verbose level of two will increase the log level to LOG_INFO which
includes logging successful mount attempts.  A log level of three or
more will add LOG_DEBUG messages and cause increasing amounts of debug
information to be logged.  The debug information exposes lots of
information about
.Nm Ns 's
inner workings which is typically only useful to developers.
Note: the
.Xr syslog 8
configuration may need to be adjusted in order to see the increased
verbosity.
.Pp
The
.Nm nfsd
utility exits 0 on success, and >0 if an error occurs.
When given the
.Cm status
command, it exits 0 if the service is enabled, and 1 if the service
is disabled.
.Pp
The following is a list of command line options that are available.
However, since
.Nm
is normally started by
.Xr launchd 8 ,
configuration of these options should be controlled using the equivalent
settings in the NFS configuration file (see
.Xr nfs.conf 5 Ns ).
.Pp
.Bl -tag -width Ds
.It Fl F Ar exports_file
The
.Ar exports_file
argument specifies an alternate location for the exports file
(useful with the
.Cm checkexports
command).
.It Fl N
Allow non-root mount requests to be served.  This should only be
specified if there are clients that require it.
.It Fl P Ar #
Specifies which port to use for the MOUNT service.
.It Fl R
Allow mount RPCs requests for regular files to be served.  Although
this seems to violate the mount protocol specification, some diskless
workstations do mount requests for their swapfiles and expect them to
be regular files.  Since a regular file cannot be specified in
.Pa /etc/exports ,
the
.Fl alldirs
option will need to be used on the export in which the swap files reside.
.It Fl n Ar #
Specifies how many NFS server threads to create.
.It Fl p Ar #
Specifies which port to use for the NFS service (instead of the default of 2049).
.It Fl r
Register the
.Tn NFS
service with
.Xr portmap 8 .
This option can be used to re-register the NFS services if the portmap
server is restarted.  This option is equivalent to the
.Cm update
command.
.It Fl t
Serve
.Tn TCP NFS
clients.  Note: only
.Tn TCP
will be used if no other options are specified.
.It Fl u
Serve
.Tn UDP NFS
clients.  Note: only
.Tn UDP
will be used if no other options are specified.
.It Fl v
Increase 
.Nm Ns 's
logging level by one (may be used multiple times).
.El
.Pp
.Sh CONFIGURATION
Unless otherwise specified, eight NFS server threads are started and both
.Tn UDP
and
.Tn TCP
transports are supported.
.Pp
A server should run enough threads to handle the maximum level of
concurrency from its clients.
.Pp
See
.Xr nfs.conf 5
for a list of tunable parameters.
.Sh FILES
.Bl -tag -width /var/run/mountdexptab -compact
.It Pa /etc/exports
The list of exported filesystems.
.It Pa /var/run/nfsd.pid
The pid of the currently running nfsd.
.It Pa /var/run/mountd.pid
The pid of the currently running mountd (provided for backwards compatibility with scripts that may HUP mountd to update exports).
.It Pa /var/run/mountdtab
The current list of outstanding mounts served.
.It Pa /var/run/mountdexptab
Information about exported file systems
and directories (UUIDs, handles, ...).
.It Pa /System/Library/LaunchDaemons/com.apple.nfsd.plist
The
.Nm
service's property list file for
.Xr launchd 8 .
.El
.Sh SEE ALSO
.Xr exports 5 ,
.Xr nfs.conf 5 ,
.Xr showmount 8 ,
.Xr nfsstat 1 ,
.Xr nfssvc 2 ,
.Xr portmap 8 ,
.Xr rpc.rquotad 8 ,
.Xr launchd 8
.Sh HISTORY
The
.Nm
and
.Cm mountd
utilities first appeared in 4.4BSD.
.Cm mountd
functionality was merged into
.Nm
in Darwin 9.


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


NFS/nfsd/nfsd.c
/*
 * Copyright (c) 1999-2007 Apple Inc.  All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Rick Macklem at The University of Guelph.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/param.h>
#include <sys/syslog.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/uio.h>
#include <sys/mount.h>
#include <sys/socket.h>
#include <sys/socketvar.h>

#include <rpc/rpc.h>
#include <rpc/pmap_clnt.h>
#include <rpc/pmap_prot.h>

#include <nfs/rpcv2.h>
#include <nfs/nfsproto.h>
#include <nfs/nfs.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <strings.h>
#include <unistd.h>
#include <pthread.h>

#include "common.h"

int nfstcpsock;

/*
 * The incredibly complex nfsd thread function
 */
static void *
nfsd_server_thread(__unused void *arg)
{
	set_thread_sigmask();
	if (nfssvc(NFSSVC_NFSD, NULL) < 0)
		log(LOG_ERR, "nfssvc: %s (%d)", strerror(errno), errno);
	DEBUG(1, "nfsd thread exiting.");
	return (NULL);
}

void *
nfsd_accept_thread(__unused void *arg)
{
	struct nfsd_args nfsdargs;
	fd_set ready, sockbits;
	struct sockaddr_in inetpeer;
	socklen_t len;
	int msgsock, on = 1;
	struct timeval tv;

	set_thread_sigmask();

	FD_ZERO(&sockbits);
	if (config.tcp)
		FD_SET(nfstcpsock, &sockbits);

	/*
	 * Loop until terminated.
	 * If accepting connections, pass new sockets into the kernel.
	 * Otherwise, just sleep waiting for a signal.
	 */
	while (!gotterm) {
		tv.tv_sec = 3600;
		tv.tv_usec = 0;
		ready = sockbits;
		if (select(nfstcpsock+1, ((nfstcpsock < 0) ? NULL : &ready), NULL, NULL, &tv) < 0) {
			if (errno == EINTR)
				continue;
			log(LOG_ERR, "select failed: %s (%d)", strerror(errno), errno);
			break;
		}
		if (config.tcp && FD_ISSET(nfstcpsock, &ready)) {
			len = sizeof(inetpeer);
			if ((msgsock = accept(nfstcpsock, (struct sockaddr *)&inetpeer, &len)) < 0) {
				log(LOG_WARNING, "accept failed: %s (%d)", strerror(errno), errno);
				continue;
			}
			DEBUG(1, "NFS socket accepted");
			memset(inetpeer.sin_zero, 0, sizeof(inetpeer.sin_zero));
			if (setsockopt(msgsock, SOL_SOCKET,
			    SO_KEEPALIVE, (char *)&on, sizeof(on)) < 0)
				log(LOG_NOTICE, "setsockopt SO_KEEPALIVE: %s (%d)", strerror(errno), errno);
			nfsdargs.sock = msgsock;
			nfsdargs.name = (caddr_t)&inetpeer;
			nfsdargs.namelen = sizeof(inetpeer);
			nfssvc(NFSSVC_ADDSOCK, &nfsdargs);
			close(msgsock);
		}
	}

	return (NULL);
}

/*
 * start a given number of NFS server threads
 */
void
nfsd_start_server_threads(int count)
{
	int threadcnt, i, rv;
	pthread_t thd;

	/* set up the server threads */
	threadcnt = 0;
	for (i=0; i < count; i++) {
		rv = pthread_create(&thd, &pattr, nfsd_server_thread, NULL);
		if (rv) {
			log(LOG_ERR, "pthread_create: %s (%d)", strerror(rv), rv);
			continue;
		}
		threadcnt++;
	}
	DEBUG(1, "Started %d of %d new nfsd threads", threadcnt, count);
	/* if no threads started exit */
	if (!threadcnt)
		log(LOG_ERR, "unable to start any nfsd threads");
	if (threadcnt != count)
		log(LOG_WARNING, "only able to create %d of %d nfsd threads", threadcnt, count);
}

/*
 * NFS server daemon mostly just a user context for nfssvc()
 *
 * 1 - set up server socket(s)
 * 2 - create the nfsd server threads
 * 3 - create the nfsd accept thread
 *
 * For connectionless protocols, just pass the socket into the kernel via.  nfssvc().
 * For connection based sockets, the accept thread loops doing the accepts.
 * When we get a new socket from accept, pass it into the kernel via. nfssvc().
 */
void
nfsd(void)
{
	struct nfsd_args nfsdargs;
	struct sockaddr_in inetaddr;
	int rv;
	int on, sock;
	pthread_t thd;

	nfstcpsock = -1;

	/* If we are serving UDP, set up the NFS/UDP socket. */
	if (config.udp) {
		if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
			log(LOG_ERR, "can't create NFS/UDP socket");
			exit(1);
		}
		inetaddr.sin_family = AF_INET;
		inetaddr.sin_addr.s_addr = INADDR_ANY;
		inetaddr.sin_port = htons(config.port);
		inetaddr.sin_len = sizeof(inetaddr);
		if (bind(sock, (struct sockaddr *)&inetaddr, sizeof(inetaddr)) < 0) {
			/* socket may still be lingering from previous incarnation */
			/* wait a few seconds and try again */
			sleep(6);
			if (bind(sock, (struct sockaddr *)&inetaddr, sizeof(inetaddr)) < 0) {
				log(LOG_ERR, "can't bind NFS/UDP addr");
				exit(1);
			}
		}
		nfsdargs.sock = sock;
		nfsdargs.name = NULL;
		nfsdargs.namelen = 0;
		if (nfssvc(NFSSVC_ADDSOCK, &nfsdargs) < 0) {
			log(LOG_ERR, "can't add NFS/UDP socket");
			exit(1);
		}
		close(sock);
	}

	/* If we are serving TCP, set up the NFS/TCP socket. */
	on = 1;
	if (config.tcp) {
		if ((nfstcpsock = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
			log(LOG_ERR, "can't create NFS/TCP socket");
			exit(1);
		}
		if (setsockopt(nfstcpsock, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(on)) < 0)
			log(LOG_WARNING, "setsockopt SO_REUSEADDR: %s (%d)", strerror(errno), errno);
		inetaddr.sin_family = AF_INET;
		inetaddr.sin_addr.s_addr = INADDR_ANY;
		inetaddr.sin_port = htons(config.port);
		inetaddr.sin_len = sizeof(inetaddr);
		if (bind(nfstcpsock, (struct sockaddr *)&inetaddr, sizeof (inetaddr)) < 0) {
			log(LOG_ERR, "can't bind NFS/TCP addr");
			exit(1);
		}
		if (listen(nfstcpsock, 128) < 0) {
			log(LOG_ERR, "NFS listen failed");
			exit(1);
		}
	}

	/* start up all the server threads */
	sysctl_set("vfs.generic.nfs.server.nfsd_thread_max", config.nfsd_threads);
	nfsd_start_server_threads(config.nfsd_threads);

	/* set up the accept thread */
	rv = pthread_create(&thd, &pattr, nfsd_accept_thread, NULL);
	if (rv) {
		log(LOG_ERR, "pthread_create: %s (%d)", strerror(rv), rv);
		exit(1);
	}
}

